src/Prefetch.cpp

```diff
@@ -3,214 +3,200 @@
+    if (stage_num == 0) {
+        return f.definition();
     }
-    return bounds;
+    return f.update(stage_num - 1);
 }
 
 class InjectPrefetch : public IRMutator {
 public:
-    InjectPrefetch(const map<string, Function> &e) : env(e) { }
+    InjectPrefetch(const map<string, Function> &e)
+        : env(e), current_func(nullptr), stage(-1) { }
 
 private:
     const map<string, Function> &env;
-    const vector<Prefetch> *prefetches = nullptr;
-    Scope<Interval> bounds;
+    const Function *current_func;
+    int stage;
+    Scope<Interval> scope;
+    Scope<Box> buffer_bounds;
 
 private:
     using IRMutator::visit;
 
-    void visit(const Let *op) {
-        Interval in = bounds_of_expr_in_scope(op->value, bounds);
-        bounds.push(op->name, in);
-        IRMutator::visit(op);
-        bounds.pop(op->name);
+    const vector<PrefetchDirective> &get_prefetch_list(const string &loop_name) {
+        if (!current_func || !starts_with(loop_name, current_func->name() + ".s" + std::to_string(stage))) {
+            vector<string> v = split_string(loop_name, ".");
+            internal_assert(v.size() > 2);
+            const string &func_name = v[0];
+
+            // Get the stage index
+            stage = -1;
+            internal_assert(v[1].substr(0, 1) == "s");
+            {
+                string str = v[1].substr(1, v[1].size() - 1);
+                bool has_only_digits = (str.find_first_not_of( "0123456789" ) == string::npos);
+                internal_assert(has_only_digits);
+                stage = atoi(str.c_str());
+            }
+            internal_assert(stage >= 0);
+
+            const auto &it = env.find(func_name);
+            internal_assert(it != env.end());
+
+            current_func = &it->second;
+            internal_assert(stage <= (int)current_func->updates().size());
+        }
+
+        const Definition &def = get_stage_definition(*current_func, stage);
+        return def.schedule().prefetches();
     }
 
-    void visit(const LetStmt *op) {
-        Interval in = bounds_of_expr_in_scope(op->value, bounds);
-        bounds.push(op->name, in);
-        IRMutator::visit(op);
-        bounds.pop(op->name);
+    const Box &get_buffer_bounds(string name, int dims) {
+        if (buffer_bounds.contains(name)) {
+            const Box &b = buffer_bounds.ref(name);
+            internal_assert((int)b.size() == dims);
+            return b;
+        }
+
+        // It is an image
+        internal_assert(env.find(name) == env.end());
+        Box b;
+        for (int i = 0; i < dims; i++) {
+            string dim_name = std::to_string(i);
+            Expr buf_min_i = Variable::make(Int(32), name + ".min." + dim_name);
+            Expr buf_extent_i = Variable::make(Int(32), name + ".extent." + dim_name);
+            Expr buf_max_i = buf_min_i + buf_extent_i - 1;
+            b.push_back(Interval(buf_min_i, buf_max_i));
+        }
+        buffer_bounds.push(name, b);
```

Comment:
I see now, the Realize visitor tracks bounds this way. I'd say that if it's not a realized buffer, that we just construct the Box and return it by value, not caching it for images.